src/storage/v2/indices/vector_index.cpp

```diff
@@ -133,142 +138,191 @@ unum::usearch::metric_kind_t VectorIndex::MetricFromName(std::string_view name)
     return unum::usearch::metric_kind_t::sorensen_k;
   }
+  if (name == "jaccard") {
+    return unum::usearch::metric_kind_t::jaccard_k;
+  }
   throw query::VectorSearchException(
       fmt::format("Unsupported metric name: {}. Supported metrics are l2sq, ip, cos, haversine, divergence, pearson, "
                   "hamming, tanimoto, and sorensen.",
                   name));
 }
 
-bool VectorIndex::CreateIndex(const VectorIndexSpec &spec, utils::SkipList<Vertex>::Accessor &vertices,
-                              std::optional<SnapshotObserverInfo> const &snapshot_info) {
-  try {
-    // Create the index
-    const unum::usearch::metric_punned_t metric(spec.dimension, spec.metric_kind, unum::usearch::scalar_kind_t::f32_k);
+utils::BasicResult<VectorIndexStorageError, VectorIndex::CreationStatus> VectorIndex::CreateIndex(
+    const VectorIndexSpec &spec, utils::SkipList<Vertex>::Accessor &vertices,
+    std::optional<SnapshotObserverInfo> const &snapshot_info) {
+  const auto label_prop = LabelPropKey{spec.label, spec.property};
+  // Create the index
+  const unum::usearch::metric_punned_t metric(spec.dimension, spec.metric_kind, unum::usearch::scalar_kind_t::f32_k);
 
-    // use the number of workers as the number of possible concurrent index operations
-    const unum::usearch::index_limits_t limits(spec.capacity, FLAGS_bolt_num_workers);
+  // use the number of workers as the number of possible concurrent index operations
+  const unum::usearch::index_limits_t limits(spec.capacity, FLAGS_bolt_num_workers);
 
-    const auto label_prop = LabelPropKey{spec.label, spec.property};
-    if (pimpl->index_.contains(label_prop) || pimpl->index_name_to_label_prop_.contains(spec.index_name)) {
-      throw query::VectorSearchException("Given vector index already exists.");
-    }
-    pimpl->index_name_to_label_prop_.try_emplace(spec.index_name, label_prop);
-    auto mg_vector_index = mg_vector_index_t::make(metric);
-    if (!mg_vector_index) {
-      throw query::VectorSearchException(fmt::format("Failed to create vector index {}, error message: {}",
-                                                     spec.index_name, mg_vector_index.error.what()));
+  if (pimpl->index_.contains(spec.index_name)) {
+    return VectorIndex::CreationStatus::ALREADY_EXISTS;
+  }
+
+  auto mg_vector_index = mg_vector_index_t::make(metric);
+  if (!mg_vector_index) {
+    spdlog::error(fmt::format("Failed to create vector index {}, error message: {}", spec.index_name,
+                              mg_vector_index.error.what()));
+    return VectorIndexStorageError::FailedToCreateIndex;
+  }
+
+  if (mg_vector_index.index.try_reserve(limits)) {
+    spdlog::info("Created vector index {}", spec.index_name);
+  } else {
+    spdlog::error(fmt::format("Unable to reserve memory for vector index {}", spec.index_name));
+    return VectorIndexStorageError::UnableToReserveMemory;
+  }
+
+  pimpl->index_name_to_label_prop_.try_emplace(spec.index_name, label_prop);
```

Comment:
Check if emplaced successfully

Reply:
emplace gives false if the key entry already exists. but we do check for that before. So I wouldn't say we need to check it?